openwisp_firmware_upgrader/upgraders/openwrt.py

```diff
@@ -155,8 +155,33 @@ def upload(self, image_file, remote_path):
         except Exception as e:
             raise RecoverableFailure(str(e))
 
+    def _verify_device_uuid(self):
+        """
+        Verifies that the UUID of the device being upgraded matches
+        the UUID in the device's configuration
+        """
+        device_uuid = self.upgrade_operation.device.pk
+        # Get UUID from device's openwisp config
+        output, exit_code = self.exec_command(
+            'uci get openwisp.http.uuid', exit_codes=[0, 1]
+        )
+        if exit_code == 1 or not output.strip():
+            self.log(_('Could not read device UUID from configuration'))
+            raise UpgradeAborted()
+        device_config_uuid = output.strip()
+        if str(device_uuid) != device_config_uuid:
+            self.log(
+                _(
+                    'Device UUID mismatch: expected {expected}, '
+                    'found {found} in device configuration'
+                ).format(expected=device_uuid, found=device_config_uuid)
+            )
+            raise UpgradeAborted()
+        self.log(_('Device identity verified successfully'))
```

Comment:
Add an assertion for the presence of this too, I had forgot to ask you this. Should be quick though.
In general, we should always assert for the presence of important UI elements we are implementing, to protect us from unwanted changes that could remove these.